### DISCUSSION TITLE
NODE_ENV is always production

### DISCUSSION THREAD

**Discussion by mo-alaa** [vercel/next.js#48914] [Help] [answered]
### Summary

Hi, when I used Docker for deployment, I noticed that NODE_ENV is always set to production. whether if we are building for development or production. Is there anyway to change it in Docker, or a custom env variable should be implemented for that case ?  

Theis happens in the third stage of the multi-stage build

I tired to pass ARG variables to the NODE_ENV but it didn't change it.

https://github.com/vercel/next.js/tree/canary/examples/with-docker
https://github.com/vercel/next.js/tree/canary/examples/with-docker-multi-env 

### Example

https://github.com/vercel/next.js/tree/canary/examples/with-docker-multi-env

**Comment by icyJoseph** [comment]
> whether if we are building for development or production.

Could you expand on what do you mean here? Which commands exactly?

**Reply by mo-alaa** [reply]
I use a Dockerfile similar to this one https://github.com/vercel/next.js/blob/canary/examples/with-docker/Dockerfile 

I tried to set NODE_ENV as follows: 
.
.
`ARG NODE_ENV`
`ENV NODE_ENV=$NODE_ENV`
`CMD ["node", "server.js"]`

Then in the code, I check `process.env.NODE_ENV` and it's always `production`

**Reply by icyJoseph** [reply, marked as answer]
I don't think one can truly ever change `NODE_ENV` like that, you'll have to use a different variable to stage/dev/integration etc, specially for the build step, for your own checks. During build or start, NODE_ENV will always be `production`.

It has been like that for a very long time, not only in Next but other frameworks. 

There's a lot of code that's conditionally added/imported/executed based of that variable, so changing it when running a build, with development is bound to cause some funky behaviour, if not impossible at all. https://overreacted.io/how-does-the-development-mode-work/

**Reply by mo-alaa** [reply]
Thanks @icyJoseph!

**Comment by rohan-charge** [comment]
Thanks a lot @icyJoseph

**Comment by stevenwoolston** [comment]
Try adding the environment variable in the CMD line. You can prefix your node scripts with variables.

For example:
`NODE_ENV=development node server.js`

**Reply by icyJoseph** [reply]
No that's not the problem.

Even there, Next might override it. Though in Next 15 one can do experimental builds with dev mode on.

**Comment by icyJoseph** [comment]
Good news everyone, as of 2024 and Next 15, - you can try the experimental feature to make a build, but with dev mode (NODE_ENV) set, PR:

- https://github.com/vercel/next.js/pull/65463
